File: packages/flutter/lib/core/domain/enum/base_enum.dart
import 'package:allowance_questboard/core/domain/value_object/base_id.dart';
import 'package:allowance_questboard/core/domain/enum/value_object/base_enum_value.dart';

/// 列挙型の基底クラス
/// 
/// 全てのEnumクラスはこのクラスを継承して実装します。
/// DartのenumではなくクラスベースのEnumシステムを提供します。
abstract class BaseEnum<T extends BaseId> {
  /// Enum値
  final BaseEnumValue<T> value;

  /// BaseEnumのコンストラクタ
  /// 
  /// [value] Enum値オブジェクト
  const BaseEnum(this.value);

  /// IDからEnum値を取得
  /// 
  /// [id] 検索するID
  /// [enumValues] 検索対象のEnum値リスト
  /// Returns: 該当するEnum値
  /// Throws: [ArgumentError] IDが見つからない場合
  static T fromId<T extends BaseEnum>(BaseId id, List<T> enumValues) {
    for (final enumValue in enumValues) {
      if (enumValue.value.id == id) {
        return enumValue;
      }
    }
    throw ArgumentError('ID $id is not valid for ${T.toString()}');
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is BaseEnum<T> && other.value == value;
  }

  @override
  int get hashCode => value.hashCode;

  @override
  String toString() => '$runtimeType(${value.toString()})';
}

File: packages/api/aqapi/core/enum/base_enum.py
from typing import TYPE_CHECKING, TypeVar, Any
from enum import Enum

from aqapi.core.domain.value_object.base_id import BaseId

IdType = TypeVar('IdType', bound=BaseId)

class BaseEnum(Enum):
    """列挙型の基底クラス"""
    
    @classmethod
    def from_id(cls, id: BaseId) -> 'BaseEnum':
        """IDから列挙型の値を取得"""
        for item in cls:
            if item.value.id == id:
                return item
        raise ValueError(f"ID {id} is not valid for {cls.__name__}")
    
